tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -127,163 +128,207 @@ public void resolve() {
+          } else {
+            throw new DAGRequestException("columns other than primary key and index key exist in fields while index single read: " + col.getName());
+          }
+        }
+        // pk is not included in index but still needed
+        if (pkIsNeeded) {
+          indexScanBuilder.addColumns(handleColumn());
+        }
+      }
+      executorBuilder.setTp(ExecType.TypeIndexScan);
 
-    Map<ColumnRef, Integer> colOffsetMap = new HashMap<>();
-    // Step1. Add columns to first executor
-    for (int i = 0; i < getFields().size(); i++) {
-      ColumnRef col = getFields().get(i);
-      tblScanBuilder.addColumns(col.getColumnInfo().toProto(tableInfo));
-      colOffsetMap.put(col, i);
-    }
-    executorBuilder.setTp(ExecType.TypeTableScan);
-    tblScanBuilder.setTableId(tableInfo.getId());
-    // Currently, according to TiKV's implementation, if handle
-    // is needed, we should add an extra column with an ID of -1
-    // to the TableScan executor
-    if (isHandleNeeded()) {
-      ColumnInfo handleColumn = ColumnInfo.newBuilder()
-          .setColumnId(-1)
-          .setPkHandle(true)
-          // We haven't changed the field name in protobuf file, but
-          // we need to set this to true in order to retrieve the handle,
-          // so the name 'setPkHandle' may sounds strange.
-          .build();
-      tblScanBuilder.addColumns(handleColumn);
-    }
-    dagRequestBuilder.addExecutors(executorBuilder.setTblScan(tblScanBuilder));
-    executorBuilder.clear();
-
-    // Step2. Add others
-    // DO NOT EDIT EXPRESSION CONSTRUCTION ORDER
-    // Or make sure the construction order is below:
-    // TableScan/IndexScan > Selection > Aggregation > TopN/Limit
-    Expression whereExpr = mergeCNFExpressions(getFilters());
-    if (whereExpr != null) {
-      executorBuilder.setTp(ExecType.TypeSelection);
-      dagRequestBuilder.addExecutors(
-          executorBuilder.setSelection(
-              Selection.newBuilder().addConditions(ProtoConverter.toProto(whereExpr, colOffsetMap))
-          )
-      );
-      executorBuilder.clear();
-    }
+      indexScanBuilder
+          .setTableId(tableInfo.getId())
+          .setIndexId(indexInfo.getId());
+      dagRequestBuilder.addExecutors(executorBuilder.setIdxScan(indexScanBuilder).build());
+    } else {
+      // TableScan
+      executorBuilder.setTp(ExecType.TypeTableScan);
+      tblScanBuilder.setTableId(tableInfo.getId());
+      // Step1. Add columns to first executor
+      for (int i = 0; i < getFields().size(); i++) {
+        ColumnRef col = getFields().get(i);
+        tblScanBuilder.addColumns(col.getColumnInfo().toProto(tableInfo));
+        colOffsetMap.put(col, i);
+      }
+      // Currently, according to TiKV's implementation, if handle
+      // is needed, we should add an extra column with an ID of -1
+      // to the TableScan executor
+      if (isHandleNeeded()) {
+        ColumnInfo handleColumn = ColumnInfo.newBuilder()
```

Comment:
why not using previously defined handleColumn?

Reply:
done